### PR TITLE
Enable Github Actions workflow to build Ubuntu-gcc 9 + MacOS Clang automatically + Other fixes

### PR DESCRIPTION
**Pull request by robromano**
This PR adds a "Build" GitHub action which builds budgetwarrior on MacOS/clang and Ubuntu/gcc9 automatically for every push.  

This PR also as a result adds support for CMake natively for budgetwarrior.  (It is probably worth considering removing Makefile and updating the docs in another PR if this is desired.)

Example successful build:
https://github.com/robromano/budgetwarrior/runs/7662965288?check_suite_focus=true

To make this work, I also noticed we had forgotten to update `cpp_utils` after adding the lambda fix from that PR. This pulls in the latest submodule to pick up this fix.  Unfortunately, it looks like the macros `HAS_MEM_FUNC` and `HAS_STATIC_FIELD` were removed from `tmp.hpp` in cpp_utils, so I decided to simply cut & paste the 2 needed macros into budgetwarrior directly.

Helps maintain fix for Issue #24

### CODE REVIEW

**Comment by wichtounet** [comment]
Hi, 

Thanks for the changes!

I can't remember why I would have removed the two macros. Probably overzealous. I have restored them in cpp_utils and upgraded cpp_utils in budgetwarrior. Can you remove this part from your PR?

> It is probably worth considering removing Makefile and updating the docs in another PR if this is desired

I won't move away from the make files, no. I much prefer it over CMake. It's fine having the Cmake configuration for Mac OS, but I will keep using the make files on Linux.

**Comment by robromano** [comment]
> I can't remember why I would have removed the two macros. Probably overzealous. I have restored them in cpp_utils and upgraded cpp_utils in budgetwarrior. Can you remove this part from your PR?

Done.

> > It is probably worth considering removing Makefile and updating the docs in another PR if this is desired
> 
> I won't move away from the make files, no. I much prefer it over CMake. It's fine having the Cmake configuration for Mac OS, but I will keep using the make files on Linux.

Sure sounds good.  It'll be there for whoever wants it, and for the build pipeline on Github Actions.

File: src/budget.cpp [hunk 1]
```diff
@@ -20,6 +20,7 @@
 #include "share.hpp"
 #include "logging.hpp"
 #include "data.hpp"
+#include "utils.hpp"
```
Comment (wichtounet) [inline src/budget.cpp:23]: Can probably remove that include now, right?

File: .github/workflows/cmake.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    - name: Build
+      # Build your program with the given configuration
+      run: cmake --build ${{github.workspace}}/build --config ${{env.BUILD_TYPE}}
+
+    - name: Test
+      working-directory: ${{github.workspace}}/build
+      # Execute tests defined by the CMake configuration.  
+      # See https://cmake.org/cmake/help/latest/manual/ctest.1.html for more detail
+      run: ctest -C ${{env.BUILD_TYPE}}
```
Comment (wichtounet) [inline .github/workflows/cmake.yml:64]: While this run, anything since tests are not in the CMake file?

File: .github/workflows/cmake.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    # The CMake configure and build commands are platform agnostic and should work equally well on Windows or Mac.
+    # You can convert this to a matrix build if you need cross-platform coverage.
+    # See: https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/managing-complex-workflows#using-a-build-matrix
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest]
+        compiler: [g++-9, clang]
+        exclude:
+          # Only default compiler on macos-latest and windows-latest
```
Comment (wichtounet) [inline .github/workflows/cmake.yml:25]:
```suggestion
          # Only default compiler on macos-latest and ubuntu-latest
```

**Review by wichtounet — APPROVED** [review]
Looks good, only minor comments.

File: .github/workflows/cmake.yml (hunk 2 above)
Comment (robromano) [reply .github/workflows/cmake.yml:64]: Correct. It will not end up running anything but it's a placeholder for when/if we add any tests.

**Pull request merged**